File: src/SmsSyncAndroid/app/src/main/java/com/example/smssyncandroid/receivers/SmsReceiver.kt
package com.example.smssyncandroid.receivers

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.provider.Telephony
import android.util.Log
import com.example.smssyncandroid.services.SmsSyncService

class SmsReceiver : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        if (intent.action == Telephony.Sms.Intents.SMS_RECEIVED_ACTION) {
            val messages = Telephony.Sms.Intents.getMessagesFromIntent(intent)
            for (sms in messages) {
                val address = sms.displayOriginatingAddress ?: continue
                val body = sms.displayMessageBody ?: ""
                val date = sms.timestampMillis

                Log.d("SmsReceiver", "SMS reçu de $address : $body")

                // Transmettre le SMS au service de synchronisation
                val serviceIntent = Intent(context, SmsSyncService::class.java).apply {
                    action = SmsSyncService.ACTION_NEW_SMS
                    putExtra(SmsSyncService.EXTRA_SMS_ADDRESS, address)
                    putExtra(SmsSyncService.EXTRA_SMS_BODY, body)
                    putExtra(SmsSyncService.EXTRA_SMS_DATE, date)
                }

                try {
                    context.startForegroundService(serviceIntent)
                } catch (e: Exception) {
                    Log.e("SmsReceiver", "Impossible de démarrer le service en premier plan: ${e.message}")
                    try {
                        context.startService(serviceIntent)
                    } catch (ex: Exception) {
                        Log.e("SmsReceiver", "Échec total du démarrage du service: ${ex.message}")
                    }
                }
            }
        }
    }
}

File: src/MobileSmsSync/Platforms/Android/Receivers/SmsReceiver.cs
using System;
using Android.App;
using Android.Content;
using Android.Provider;
using Android.Util;

namespace MobileSmsSync.Services
{
    [BroadcastReceiver(Name = "com.companyname.mobilesmssync.SmsReceiver", Enabled = true, Exported = true)]
    [IntentFilter(new[] { "android.provider.Telephony.SMS_RECEIVED" }, Priority = 999)]
    public class SmsReceiver : BroadcastReceiver
    {
        public override void OnReceive(Context? context, Intent? intent)
        {
            if (context == null || intent == null) return;

            if (intent.Action == "android.provider.Telephony.SMS_RECEIVED")
            {
                try
                {
                    var messages = Telephony.Sms.Intents.GetMessagesFromIntent(intent);
                    if (messages == null) return;

                    foreach (var sms in messages)
                    {
                        if (sms == null) continue;

                        string address = sms.DisplayOriginatingAddress ?? string.Empty;
                        string body = sms.DisplayMessageBody ?? string.Empty;
                        long date = sms.TimestampMillis;

                        Log.Debug("SmsReceiver", $"SMS reçu de {address} : {body}");

                        if (!string.IsNullOrEmpty(address))
                        {
                            var serviceIntent = new Intent(context, typeof(SmsSyncService));
                            serviceIntent.SetAction(SmsSyncService.ActionNewSms);
                            serviceIntent.PutExtra(SmsSyncService.ExtraSmsAddress, address);
                            serviceIntent.PutExtra(SmsSyncService.ExtraSmsBody, body);
                            serviceIntent.PutExtra(SmsSyncService.ExtraSmsDate, date);

                            if (Android.OS.Build.VERSION.SdkInt >= Android.OS.BuildVersionCodes.O)
                            {
                                context.StartForegroundService(serviceIntent);
                            }
                            else
                            {
                                context.StartService(serviceIntent);
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    Log.Error("SmsReceiver", $"Erreur réception SMS: {ex.Message}");
                }
            }
        }
    }
}
